src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -599,7 +602,20 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
                         index_and_condition.first, index_and_condition.second, part, ranges.ranges, settings, reader_settings, log);
 
             if (!ranges.ranges.empty())
+            {
+                if (settings.read_overflow_mode == OverflowMode::THROW && settings.max_rows_to_read)
+                {
+                    /// Fail fast if estimated number of rows to read exceeds the limit
+                    size_t total_rows_estimate = total_rows.fetch_add(ranges.getRowsCount());
```

Comment:
`fetch_add` returns the value that was there before the addition. You can make it fail a little bit earlier if you compare the estimate after incrementing.

```cpp
auto part_rows_estimate = ranges.getRowsCount();
size_t prev_total_rows_estimate = total_rows.fetch_add(parts_rows_estimate);
if (prev_total_rows_estimate + part_rows_estimate > settings.max_rows_to_read) { ... }
```